client/client-multi/client-c/src/kaa/platform-impl/posix/key_utils.c

```diff
@@ -0,0 +1,417 @@
+    }
+
+    /* Adding PKCS7 padding */
+    size_t enc_data_size = ext_get_encrypted_data_size(payload_size);
+    uint8_t padding = enc_data_size - payload_size;
+    memset(output + payload_size, padding, padding);
+
+    /* Process the buffer chunk by chunk */
+    while (enc_data_size >= KAA_SESSION_KEY_LENGTH) {
+        if (aes_encrypt_decrypt(MBEDTLS_AES_ENCRYPT, input,
+                KAA_SESSION_KEY_LENGTH,
+                output, keys.session_key) < 0) {
+            return KAA_ERR_BADDATA;
+        }
+
+        enc_data_size -= KAA_SESSION_KEY_LENGTH;
+
+        input += KAA_SESSION_KEY_LENGTH;
+        output += KAA_SESSION_KEY_LENGTH;
+    }
+
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t ext_decrypt_data(const uint8_t *input, size_t input_size,
+        uint8_t *output, size_t *output_payload_size)
+{
+    if (!input || !input_size || !output || !output_payload_size) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    size_t tmp_in_size = input_size;
+    while (input_size >= KAA_SESSION_KEY_LENGTH) {
+        if (aes_encrypt_decrypt(MBEDTLS_AES_DECRYPT, input,
+                KAA_SESSION_KEY_LENGTH, output,
+                keys.session_key) < 0) {
+            return KAA_ERR_BADDATA;
+        }
+        input_size -= KAA_SESSION_KEY_LENGTH;
+
+        input += KAA_SESSION_KEY_LENGTH;
+        output += KAA_SESSION_KEY_LENGTH;
+    }
+
+    /* Reduce PKCS7 padding */
+    uint8_t padding_length = *(output - 1);
+    *output_payload_size = tmp_in_size - padding_length;
+
+    return KAA_ERR_NONE;
+}
+
+/* Get encrypted session key (enctypted with remote key) */
+kaa_error_t ext_get_encrypted_session_key(uint8_t **buffer, size_t *buffer_size,
+        const uint8_t *remote_key, size_t remote_key_size)
+{
+    if (!buffer || !buffer_size || !remote_key || !remote_key_size) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    mbedtls_pk_context local_ctx;
+    mbedtls_pk_init(&local_ctx);
+    int err = set_rsa_public_key(&local_ctx, remote_key, remote_key_size);
+    if (err) {
+        goto exit;
+    }
+    err = rsa_encrypt(&local_ctx, keys.session_key,
+            KAA_SESSION_KEY_LENGTH, keys.encrypted_session_key);
+    if (err) {
+        goto exit;
+    }
+
+    keys.encrypted_session_key_length = ENC_SESSION_KEY_LENGTH;
+
+    *buffer = keys.encrypted_session_key;
+    *buffer_size = keys.encrypted_session_key_length;
+
+exit:
+    mbedtls_pk_free(&local_ctx);
+    return err ? KAA_ERR_BADDATA : KAA_ERR_NONE;
+}
+
+size_t ext_get_encrypted_data_size(size_t input_size)
+{
+    if (!input_size) {
+        return 0;
+    }
+
+    if (input_size % AES_ECB_ENCRYPTION_CHUNCK_SIZE != 0) {
+        return AES_ECB_ENCRYPTION_CHUNCK_SIZE -
+               (input_size % AES_ECB_ENCRYPTION_CHUNCK_SIZE) + input_size;
+    }
+
+    return input_size + AES_ECB_ENCRYPTION_CHUNCK_SIZE;
+}
+
+static int rsa_sign(mbedtls_pk_context *pk, const uint8_t *input,
+        size_t input_size, uint8_t *output, size_t *output_size)
+{
```

Comment:
Arguments worth checking. Return error and place TODO here: http://jira.kaaproject.org/browse/KAA-982